jane/doc/parallelism-tutorial.md

```diff
@@ -1,38 +1,162 @@
+`initialised` to `true`. Or maybe domain A sets both variables before domain B
+checks `initialised`, and then if gold is cheap, domain B heads to the market.
+
+The one thing we want _not_ to happen is for domain B to see that
+`!initialised` is `true` and yet find that `price_of_gold` is still stuck at
+its initial value of `0.0`, thus causing us to `buy_lots_of_gold ()` no matter
+the price. Our intuitive method reassures us that this cannot happen: domain A
+doesn't set `initialised` to `true` until it's already done setting
+`price_of_gold` to something sensible.
+
+Unfortunately, our intuitive reasoning is wrong: this code very much can end up
+calling `buy_lots_of_gold ()` despite `price_of_gold` being high. Since the
+assignments to `price_of_gold` and `initialised` happen in the same domain and
+neither is a fancy [atomic] operation, they can be reordered so that
+`initialised` is set first, _while `price_of_gold` is still zero,_ leading
+domain B to make an expensive mistake.
+
+[atomic]: #atomics
+
+It turns out our intuition assumes what the literature calls
+_sequential consistency:_ essentially, the principle that things happen in order
+they're written, and parallelism just means we're not sure which domain will do
+its next thing. This principle is so essential that it's hard to imagine
+reasoning about a multicore program without it.
+
+Of course, what's happened here is two data races: each of `price_of_gold` and
+`initialised` is being modified by domain A and read by domain B in parallel.
+So the moral of the story is that data races violate sequential consistency.
+
+The good news is that the inverse is also true: any program without data races
+is sequentially consistent. Data-race freedom gives us back the power to reason
+intuitively about our code, no matter how buggy it might get. In our example,
+that means that once we fix the data races (easy to do with `Atomic.t`), we know
+that fixes our bug: sequential consistency means that it's impossible for domain
+B to observe a state where `initialised` is `true` and `price_of_gold` is
+`0.0`.[^price-zero]
+
+[^price-zero]: Assuming, of course, `calculate_price_of_gold ()` doesn't
+return `0.0`. The compiler can't catch everything.
+
+The even _better_ news is that, in code using OxCaml's extensions, the data
+races in `price_of_gold` and `initialised` would _never have compiled to begin
+with._ There's a tradeoff, of course: just as OCaml's type safety means having
+to understand the type system, the system for data-race freedom has a learning
+curve. This tutorial aims to get you moving with a practical framework that
+hopes to achieve significant parallel speedups without much fuss. You'll still
+need to convince the compiler that your code has no data races, but we'll walk
+through the rules and even some [shortcuts](#amenities) for common special
+cases.
+
+# Fork/join parallelism
+
+Any problem that can be neatly subdivided is a good candidate for
```

Comment:
"subdivided into parallel pieces"? But that makes the sentence a little circular. I mean, *any* problem can be subdivided. So I think this isn't exactly what you mean (but fail to have a better phrasing at the moment).

Reply:
Yeah, I struggled a bit too. The word _neatly_ is sort of standing in for the idea that the pieces have to be separate but it's not very precise.